docs/dev/alternative_release_image_sources.md

```diff
@@ -0,0 +1,154 @@
+
+`release-image` content is what operators get installed and what version of each operator (i.e. the container image) gets installed to the cluster. While, `release-image` source is **where** the `release-image` content gets pulled.
+
+Currently the installer controls **both** the `release-image` content and the `release-image` source using the embedded release-image location or the `OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE` env, but to allow users to use a release-image from their private registry (change the `release-image` source), but keep the `release-image` content identical to one vetted by OpenShift, the installer needs to allow specifying source separate from the content.
+
+## Controlling the content
+
+The content of the `release-image`, i.e. the digest, continues to the controlled by the embedded release-image location or the `OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE` env.
+
+## Controlling the source
+
+The installer allows the users to specify the sources for the release-image repository and the other repositories referenced in the release-image through the InstallConfig.
+
+## Details
+
+## InstallConfig
+
+```go
+type InstallConfig struct {
+    ///
+
+    ReleaseImageSources []ReleaseImageSource
+
+    ///
+}
+
+type ReleaseImageSource struct {
+    Repository string
+    // Sources is the list of repositories that can be used to fetch content from Repository.
+    Sources []ImageSource
+}
+```
+
+If the release-image `q.io/ocp/release-x.y@sha256:abc` which has references to the images in repositories `q.io/openshift/x.y` and `q.io/rhhi/x.y` was mirrored to `local.registry.com/ocp/release-x.y`, the `install-config.yaml` would look like,
+
+```yaml
+...
+releaseImageSources:
+- repository: q.io/ocp/release-x.y
+  sources:
+  - repository: local.registry.com/ocp/release-x.y
+- repository: q.io/openshift/x.y
+  sources:
+  - repository: local.registry.com/ocp/release-x.y
+- repository: q.io/rhhi/x.y
+  sources:
+  - repository: local.registry.com/ocp/release-x.y
+...
+```
+
+### ImageContentSourcePolicy
+
+If a list of `ReleaseImageSources` is specified, the installer configures the [RepositoryDigestMirrors][repository-digest-mirrors] for each `ReleaseImageSource` from the list as mirrors of each other.
+
+If the `install-config.yaml` provided by the user is:
+
+```yaml
+...
+releaseImageSources:
+- repository: q.io/ocp/release-x.y
+  sources:
+  - repository: local.registry.com/ocp/release-x.y
+- repository: q.io/openshift/x.y
+  sources:
+  - repository: local.registry.com/ocp/release-x.y
+- repository: q.io/rhhi/x.y
+  sources:
+  - repository: local.registry.com/ocp/release-x.y
+...
+```
+
+The `ImageContentSourcePolicy` object would look like:
+
+```yaml
+...
+repositoryDigestMirrors:
+- sources:
+  - q.io/ocp/release-x.y
+  - local.registry.com/ocp/release-x.y
+- sources:
+  - q.io/openshift/x.y
+  - local.registry.com/ocp/release-x.y
+- sources:
+  - q.io/rhhi/x.y
+  - local.registry.com/ocp/release-x.y
```

Comment:
If this is an ordered list, don't we want the primary repository to be listed at the end so that the user knows that the mirrors will be tried first and primary last.
Also, since `repositoryDigestMirrors` is just a list of sources, we don't really know what the primary repo is, does it mean `registries.conf` should have an entry for each source listed as the primary? Something like if `repoistoryDigestMirrors: A, B, C, D` then registries.conf will have the following entries
```
[[registry]]location = A; mirrors = B, C, D
[[registry]]location = B; mirrors = A, C, D
[[registry]]location = C; mirrors = A, B, D
[[registry]]location = D; mirrors = A, B, C
```
cc @mtrmac @mrunalp

Reply:
Actually, if `repositoryDigestMirrors` is supposed to be a consistent order, the `mirrors` list would also be the same, and include the primary one:
```
[[registry]]location = A; mirrors = A, B, C, D
[[registry]]location = B; mirrors = A, B, C, D
[[registry]]location = C; mirrors = A, B, C, D
[[registry]]location = D; mirrors = A, B, C, D
```
(and `registries.conf` would be modified not to contact the primary one again if it is already listed in mirrors.)

---

WRT this PR:
- `ReleaseImageSources` should _document_ whether `repository` or `sources.repository` is supposed to be accessed first. Assuming `repository` is the internet-accessible pull spec and `sources.repository` is the list of. mirrors, presumably the mirrors should be tried first.
- The example of generated `repositoryDigestMirrors` (and the code generating it) should generate an array of repositories in the order in which they should be attempted; Again, assuming `repository` is the internet-accessible pull spec and `sources.repository` is the list of. mirrors, and that the mirrors should be tried first, the example above should have `local.registry.com` entries before `q.io` entries in each `sources` array.

Reply:
I think in general mirrors should be preferred (we get an advantage in both cases) when specified.